src/libraries/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Android/ChainPal.cs

```diff
@@ -35,7 +38,319 @@ public static bool ReleaseSafeX509ChainHandle(IntPtr handle)
             TimeSpan timeout,
             bool disableAia)
         {
-            throw new NotImplementedException(nameof(BuildChain));
+            var chainPal = new AndroidCertPath();
+            try
+            {
+                chainPal.Initialize(cert, extraStore, customTrustStore, trustMode);
+                chainPal.Evaluate(verificationTime, applicationPolicy, certificatePolicy, revocationMode, revocationFlag);
+            }
+            catch
+            {
+                chainPal.Dispose();
+                throw;
+            }
+
+            return chainPal;
+        }
+
+        private sealed class AndroidCertPath : IChainPal
+        {
+            public X509ChainElement[]? ChainElements { get; private set; }
+            public X509ChainStatus[]? ChainStatus { get; private set; }
+
+            public SafeX509ChainHandle? SafeHandle => null;
+
+            private SafeX509ChainContextHandle? _chainContext;
+            private bool _isValid;
+
+            public void Dispose()
+            {
+                if (_chainContext != null)
+                {
+                    _chainContext.Dispose();
+                }
+            }
+
+            public bool? Verify(X509VerificationFlags flags, out Exception? exception)
+            {
+                Debug.Assert(_chainContext != null);
+                exception = null;
+
+                if (!_isValid)
+                {
+                    // There is no way to bypass time, trusted root, name, or policy constraint
+                    // validation on Android. It will not build any chain without these all
+                    // being valid.
+                    X509VerificationFlags[] unsupportedFlags = new X509VerificationFlags[]
+                    {
+                        X509VerificationFlags.IgnoreNotTimeValid,
+                        X509VerificationFlags.AllowUnknownCertificateAuthority,
+                        X509VerificationFlags.IgnoreInvalidName,
+                        X509VerificationFlags.IgnoreInvalidPolicy,
+                    };
+                    foreach (X509VerificationFlags unsupported in unsupportedFlags)
+                    {
+                        if ((flags & unsupported) == unsupported)
+                        {
+                            exception = new PlatformNotSupportedException(SR.Format(SR.Chain_VerificationFlagNotSupported, unsupported));
```

Comment:
This is explicitly doing PNSE if some ignore flags are set that Android doesn't handle and the chain couldn't be built. We don't really have a way of determining why it couldn't be built - it might not actually be for one of the things the user tried to ignore. Maybe it would make sense to just let it return false without throwing? (and document that certain verification flags aren't supported)

Reply:
Hm. Throwing here does seem weird... because "you can use this value unless it maybe mattered" is weird.

Throwing always in chain.Build when an invalid flag is specified makes sense to me.  Or, yeah, returning false and not throwing.

The behavior for any of these cases is the zero-certificate PartialChain, right?  If so, just returning false is probably sufficient.

Reply:
Yep, any of these would result in zero-certificate PartialChain. Will go with returning false.